README.md
# skilljaR

`skilljaR` is an R wrapper for the Skilljar API.

## Installation

Install from CRAN with `install.packages("skilljaR").`

Install the development version with `devtools::install_github("chrisumphlett/skilljaR")`.

## Functionality

* `get_users()` returns a data.frame with all users from a particular domain. The API only returns up to 10,000 users at a time. The function makes multiple requests to the API in order to return the desired number of users. The user needs to know, or guess, how many users they have if they want to return them all. You can select an arbitrarily large number to be safe but it will result in a longer execution time as there will be unnecessary API requests sent.
* `get_published_courses()` returns a data.frame with all published courses in a particular domain.
* `get_course_progress()` returns a data.frame with the course progress progress for user enrollments in your courses. You must supply a vector of user id's (which can be generated from the results of `get_users()`).

The motivation for this package was to replicate, and improve upon, the manual enrollments data export in the Skilljar portal. The results of `get_course_progress()` joined back together with the results of `get_users()` should reproduce that export but with additional fields. Some of the columns are named differently.

## Security
Utilizing the API requires a token. This must be obtained by logging in at dashboard.skilljar.com and going to Organization / API Credentials. There are different strategies for storing api tokens securely. This package does not require nor nudge you towards any one strategy. That said, it is an unnecessary risk to store the token in the script!

news.md
# skilljaR 0.1.2.9000

* Dev version

# skilljaR 0.1.2

* Created `get_user_domain()` which will return a data frame with requested user id and a column indicating if the user is in the domain.

# skilljaR 0.1.1

* Created `get_course_paths()` which will retrieve the list of paths and then the items (courses) associated with each path.

# skilljaR 0.1.0

* API endpoint for `get_users()` changed, to an endpoint that provides the `latest_activity` for each user. This is a significant improvement, enabling one to retrieve only updated course progress for users who have recent activity.
* This is a breaking change, as the `domains` argument is no longer required. This is an improvement -- now multiple API calls are not required.
* The users desired input is also not required anymore. By default the function will get all users. This argument would be used only to limit the of users (for testing purposes, for instance). The function will return all the pages, this argument isn't used to force it to get to enough pages either (and there is no risk of having redundant unnecessary calls).
* The `encoding_` arguments were removed for all functions, and `UTF-8` is being used by default.

# skilljaR 0.0.1

* The first version of the package.

R/get_user_domain.R
#' Determine if User has joined a domain
#' 
#' Accepts a vector of user id's, and a domain, and returns a data.frame that
#' indicates if each id is found in that domain.
#' 
#' Utilizing the API requires a token. This must be obtained by logging in
#' at dashboard.skilljar.com and going to Organization -> API Credentials.
#' There are different strategies for storing api tokens securely. It is 
#' an unnecessary risk to store the token in the script!
#' 
#' @param user_ids Vector of user id's
#' @param domain Skilljar domain you want to search in
#' @param api_token Your personalized token provided by 'Skilljar'
#' 
#' @import httr
#' @importFrom purrr "map_dfr"
#' @import jsonlite
#' 
#' @return A data.frame that indicates if each id is found in that domain.
#' 
#' @seealso See \url{https://api.skilljar.com/docs/} for documentation on
#' the 'Skilljar' API.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' # Retrieve user data and domain for 100 users
#' my_users <- get_users(users_desired = 100,
#' api_token = "my-token")
#' user_vector <- users %>% select(user_id) %>% pull()
#' my_users2 <- get_user_domain(user_vector, "my-domain.skilljar.com",
#' api_token = "my-token")
#' }

get_user_domain <- function(#users_desired = 100000000, 
  domain, user_ids, api_token){
  token <- jsonlite::base64_enc(api_token)
  
  get_user_domains <- function(.u){
    results <- httr::GET(paste0("https://api.skilljar.com/v1/domains/",
                                domain,
                                "/users/",
                                .u),
                                #?page_size=1&page=1",
                         httr::add_headers(
                           "Authorization" = paste(
                             "Basic",
                             gsub("\n", "", token)),
                           "Content-Type" = paste0(
                             "application/x-www-form-urlencoded;charset=UTF-8"
                           )),
                         type = "basic"
    )
    get_text <- httr::content(results, "text", encoding = "UTF-8")
    if (get_text == "{\"detail\":\"Not found.\"}"){
      in_domain <- FALSE
    } else {
      in_domain <- TRUE
    }
    df <- tibble(user_id = .u, in_domain = in_domain)
  }
  
  df <- purrr::map_dfr(user_ids, get_user_domains)
  return(df)
  
}


R/get_course_paths.R
#' Get Course Paths
#' 
#' Returns a data.frame with all paths and the items in those paths.
#' 
#' Utilizing the API requires a token. This must be obtained by logging in
#' at dashboard.skilljar.com and going to Organization -> API Credentials.
#' There are different strategies for storing api tokens securely. It is 
#' an unnecessary risk to store the token in the script!
#' 
#' @param api_token Your personalized token provided by 'Skilljar'
#' 
#' @import httr
#' @import dplyr
#' @importFrom magrittr "%>%"
#' @importFrom purrr "map_dfr"
#' @import jsonlite
#' 
#' @return A data frame with users and user data
#' 
#' @seealso See \url{https://api.skilljar.com/docs/} for documentation on
#' the 'Skilljar' API.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' # Get course progress data
#' paths <- get_course_paths(api_token = "my-token")
#' }

get_course_paths <- function(api_token){
  
  token <- jsonlite::base64_enc(api_token)
  
  # get the paths first
  api_path <- paste0("https://api.skilljar.com/v1/paths")
  get_results <- httr::GET(api_path, httr::add_headers(
    "Authorization" = paste("Basic", gsub("\n", "", token)),
    "Content-Type" = "application/x-www-form-urlencoded;charset=UTF-8"
  ), type = "basic")
  paths_df <- as.data.frame(
    jsonlite::fromJSON(
      httr::content(get_results, "text", encoding = "UTF-8"),
      flatten = TRUE
      )$results
    ) %>%
    rename(path_id = .data$id, path_name = .data$title)
  
  # list of path id's in a vector to pass into next call
  path_ids <- pull(paths_df, .data$path_id)

  map_across_paths <- function(.x){
    
    api_path <- paste0("https://api.skilljar.com/v1/paths/",
                       .x,
                       "/path-items")
    
    get_results <- httr::GET(api_path,
                             httr::add_headers(
                               "Authorization" = paste(
                                 "Basic",
                                 gsub("\n", "", token)),
                               "Content-Type" = paste0(
                                 "application/x-www-form-urlencoded;charset=UTF-8")),
                             type = "basic"
    )
    items_df <- as.data.frame(
      jsonlite::fromJSON(
        httr::content(get_results, "text", encoding = "UTF-8"),
        flatten = TRUE
        )$results
      ) %>%
      dplyr::mutate(path_id = .x)
  }
  all_path_items <- purrr::map_dfr(path_ids, map_across_paths) %>%
    dplyr::left_join(paths_df)
}


R/get_course_progress.R
#' Get Course Progress for a List of Users
#' 
#' Returns a data.frame with all course progress details for a list of user
#' id's. 
#' 
#' The API call only returns one user at a time. If you have many users then
#' it may take a long time for the function to return. You may also exceed the
#' 'Skilljar' API rate limit.
#' 
#' Utilizing the API requires a token. This must be obtained by logging in
#' at dashboard.skilljar.com and going to Organization -> API Credentials.
#' There are different strategies for storing api tokens securely. It is 
#' an unnecessary risk to store the token in the script!
#' 
#' @param user_ids Vector of user id's for which you want course progress data
#' @param api_token Your personalized token provided by 'Skilljar'
#' 
#' @import httr
#' @import dplyr
#' @importFrom magrittr "%>%"
#' @importFrom purrr "is_empty"
#' @importFrom purrr "map_dfr"
#' @import jsonlite
#' 
#' @return A data frame with users and user data
#' 
#' @seealso See \url{https://api.skilljar.com/docs/} for documentation on
#' the 'Skilljar' API.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' # Get some user id's
#' users <- get_users(domain = "training.mycompany.com",
#' users_desired = 10,
#' api_token = "my-token")
#' 
#' # Get course progress data
#' progress <- get_course_progress(user_ids = unique(users$user.id),
#' api_token = "my-token")
#' }

get_course_progress <- function(user_ids, api_token){
  
  token <- jsonlite::base64_enc(api_token)
  # i <- 1
  map_across_users <- function(.x){

    api_path <- paste0("https://api.skilljar.com/v1/users/",
                       .x,
                       "/published-courses")
    
    get_results <- httr::GET(api_path,
                       httr::add_headers(
                         "Authorization" = paste(
                           "Basic",
                           gsub("\n", "", token)),
                         "Content-Type" = paste0(
                           "application/x-www-form-urlencoded;charset=UTF-8")),
                       type = "basic"
    )
    get_text <- httr::content(get_results, "text", encoding = "UTF-8")
    
    ## skip this if there were no results for the user
    if(get_text != "[]"){
      user_progress <- as.data.frame(jsonlite::fromJSON(get_text,
                                                        flatten = TRUE)) %>%
        dplyr::mutate(user_id = .x) %>%
        dplyr::select(-.data$all_enrollments)
    }
    # i <- 1 + 1

  }
  all_progress <- purrr::map_dfr(user_ids, map_across_users)
}


R/get_published_courses.R
#' Get All Published Courses
#' 
#' Returns a data.frame with all published courses from a particular domain.
#' 
#' Utilizing the API requires a token. This must be obtained by logging in
#' at dashboard.skilljar.com and going to Organization -> API Credentials.
#' There are different strategies for storing api tokens securely. It is 
#' an unnecessary risk to store the token in the script!
#' 
#' @param domain Domain of 'Skilljar' account
#' @param api_token Your personalized token provided by 'Skilljar'
#' 
#' @import httr
#' @importFrom purrr "is_empty"
#' @import jsonlite
#' @import dplyr
#' 
#' @return A data frame with all published course data
#' 
#' @seealso See \url{https://api.skilljar.com/docs/} for documentation on
#' the 'Skilljar' API.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' # Retrieve all published courses
#' courses <- get_published_courses(domain = "training.mycompany.com",
#' api_token = "my-token")
#' }

get_published_courses <- function(domain, api_token){

  token <- jsonlite::base64_enc(api_token)

  api_path <- paste0(
    "https://api.skilljar.com/v1/domains/",
    domain,
    "/published-courses")
  
  keep_going <- TRUE
  
  while (keep_going == TRUE) {
    get_results <- httr::GET(api_path,
                       httr::add_headers(
                         "Authorization" = paste(
                           "Basic",
                           gsub("\n", "", token)),
                         "Content-Type" = paste0(
                           "application/x-www-form-urlencoded;charset=UTF-8")),
                       type = "basic"
    )
    get_text <- httr::content(get_results, "text", encoding = "UTF-8")
    get_json <- jsonlite::fromJSON(get_text, flatten = TRUE)
    courses_page <- as.data.frame(get_json$results)
    
    if (exists("all_published_courses")){
      all_published_courses <- dplyr::bind_rows(all_published_courses, courses_page)
    } else {
      all_published_courses <- courses_page
    }
    
    if (purrr::is_empty(unlist(get_json$`next`))){
      keep_going <- FALSE
    } else {
      api_path <- get_json$`next`
    }
  }
  return(all_published_courses)
}


R/get_users.R
#' Get All Users
#' 
#' Returns a data.frame with all users from a particular domain. If you have 
#' many users and do not want to return them all at once, you may request
#' fewer users to save time while doing development. The API returns up 10,000
#' users at a time-- if you request more than 10,000, it will return in full
#' page increments (a multiple of 10,000).
#' 
#' Utilizing the API requires a token. This must be obtained by logging in
#' at dashboard.skilljar.com and going to Organization -> API Credentials.
#' There are different strategies for storing api tokens securely. It is 
#' an unnecessary risk to store the token in the script!
#' 
#' @param users_desired Number of user records to return
#' @param api_token Your personalized token provided by 'Skilljar'
#' 
#' @import httr
#' @importFrom purrr "map_dfr"
#' @import jsonlite
#' 
#' @return A data frame with users and user data
#' 
#' @seealso See \url{https://api.skilljar.com/docs/} for documentation on
#' the 'Skilljar' API.
#' 
#' @export
#' 
#' @examples
#' \dontrun{
#' # Retrieve 1000 users
#' my_users <- get_users(users_desired = 1000,
#' api_token = "my-token")
#' }

get_users <- function(users_desired = 100000000, api_token){
  token <- jsonlite::base64_enc(api_token)
  
  get_pages <- httr::GET("https://api.skilljar.com/v1/users?page_size=1&page=1",
                         httr::add_headers(
                           "Authorization" = paste(
                             "Basic",
                             gsub("\n", "", token)),
                           "Content-Type" = paste0(
                             "application/x-www-form-urlencoded;charset=UTF-8"
                             )),
                         type = "basic"
  )
  get_text <- httr::content(get_pages, "text", encoding = "UTF-8")
  total_pages <- floor(
                  jsonlite::fromJSON(get_text, flatten = TRUE)$count / 10000
                  ) + 1
  pages_to_process <- min(total_pages, floor(users_desired / 10000) + 1)
  pages_list <- seq(from = 1, to = pages_to_process, by = 1)
  
  page_length <- min(users_desired, 10000)
  
  get_results <- function(x) {
    api_path <- paste0(
      "https://api.skilljar.com/v1/users?page_size=",
      page_length,
      "&page=",
      x)

    api_return <- httr::GET(api_path,
                           httr::add_headers(
                             "Authorization" = paste(
                               "Basic",
                               gsub("\n", "", token)),
                             "Content-Type" = paste0(
                               "application/x-www-form-urlencoded;charset=UTF-8"
                             )),
                           type = "basic"
                           )
    get_text <- httr::content(api_return, "text", encoding = "UTF-8")
    get_json <- jsonlite::fromJSON(get_text, flatten = TRUE)
    users_df <- as.data.frame(get_json$results)
    return(users_df)
  }
  
  df <- purrr::map_dfr(pages_list, get_results)
  
  return(df)

}
